src/test/java/teammates/common/datatransfer/questions/FeedbackContributionQuestionDetailsTest.java

```diff
@@ -21,4 +32,173 @@ public void testIsFeedbackParticipantCommentsOnResponsesAllowed_shouldReturnFals
+        ______TS("failure: mix of valid and invalid responses");
+        responses.clear();
+        expectedResponsesValidationResults.clear();
+        for (int answer : VALID_CONTRIBUTION_RESPONSE_ANSWERS) {
+            FeedbackContributionResponseDetails details = new FeedbackContributionResponseDetails();
+            details.setAnswer(answer);
+            responses.add(details);
+        }
+        for (int answer : INVALID_CONTRIBUTION_RESPONSE_ANSWERS) {
+            FeedbackContributionResponseDetails details = new FeedbackContributionResponseDetails();
+            details.setAnswer(answer);
+            responses.add(details);
+            expectedResponsesValidationResults.add(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_OPTION);
+        }
+        assertEquals(expectedResponsesValidationResults,
+                feedbackContributionQuestionDetails.validateResponsesDetails(responses, 0));
+
+    }
+
+    @Test
+    public void testValidateGiverRecipientVisibility() {
+        FeedbackContributionQuestionDetails details = new FeedbackContributionQuestionDetails();
+        FeedbackQuestionAttributes feedbackQuestionAttributes = FeedbackQuestionAttributes.builder()
+                .withCourseId("course")
+                .withFeedbackSessionName("session")
+                .withGiverType(FeedbackParticipantType.STUDENTS)
+                .withRecipientType(FeedbackParticipantType.OWN_TEAM_MEMBERS_INCLUDING_SELF)
+                .withQuestionNumber(1)
+                .withNumberOfEntitiesToGiveFeedbackTo(Const.MAX_POSSIBLE_RECIPIENTS)
+                .withShowResponsesTo(Arrays.asList(FeedbackParticipantType.RECEIVER,
+                        FeedbackParticipantType.RECEIVER_TEAM_MEMBERS,
+                        FeedbackParticipantType.OWN_TEAM_MEMBERS,
+                        FeedbackParticipantType.INSTRUCTORS))
+                .withShowGiverNameTo(new ArrayList<>())
+                .withShowRecipientNameTo(new ArrayList<>())
+                .withQuestionDescription("description")
+                .build();
+
+        ______TS("success: valid giver recipient visibility");
+        assertEquals("", details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
+        ______TS("failure: giver type is not STUDENT");
+        feedbackQuestionAttributes.setGiverType(FeedbackParticipantType.SELF);
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_FEEDBACK_PATH,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
+        ______TS("failure: recipient type can only be OWN_TEAM_MEMBERS_INCLUDING_SELF");
+        feedbackQuestionAttributes.setRecipientType(FeedbackParticipantType.SELF);
+        assertEquals(FeedbackContributionQuestionDetails.CONTRIB_ERROR_INVALID_FEEDBACK_PATH,
+                details.validateGiverRecipientVisibility(feedbackQuestionAttributes));
+
+        ______TS("failure: invalid restrictions on visibility options");
```

Comment:
We can add more scenarios here - we only tested a little

Reply:
Fixed